### PR TITLE
Pms v15 mn tests

### PR DESCRIPTION
**Pull request by frederikgeth**
936 passed, 104 failed, 62 errored, 0 broken

### CODE REVIEW

File: src/core/variable.jl [hunk 1]
```diff
@@ -462,107 +469,131 @@ Create a dictionary with values of type Any for the load.
...
     end
+
+    report && _PMs.sol_component_value(pm, nw, :storage, :status, _PMs.ids(pm, nw, :storage), z_storage)
 end
 
 
 "Create variables for `active` and `reactive` storage injection"
 function variable_mc_on_off_storage(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, kwargs...)
-    variabe_mc_on_off_storage_active(pm; nw=nw, kwargs...)
+    variable_mc_on_off_storage_active(pm; nw=nw, kwargs...)
```
Comment (sanderclaeys) [inline src/core/variable.jl]: I don't see any difference here

File: src/core/variable.jl [hunk 2]
```diff
@@ -462,107 +469,131 @@ Create a dictionary with values of type Any for the load.
...
     variable_mc_on_off_storage_reactive(pm; nw=nw, kwargs...)
 end
 
 
 "Create variables for `active` storage injection"
-function variabe_mc_on_off_storage_active(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw)
+function variable_mc_on_off_storage_active(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, report::Bool=true)
+    cnds = _PMs.conductor_ids(pm; nw=nw)
```
Comment (sanderclaeys) [inline src/core/variable.jl]: cnds is not used

File: src/core/variable.jl [hunk 3]
```diff
@@ -571,51 +602,63 @@ function variabe_mc_on_off_storage_active(pm::_PMs.AbstractPowerModel; nw::Int=p
...
     ) for i in _PMs.ids(pm, nw, :storage))
+
+    report && _PMs.sol_component_value(pm, nw, :storage, :ps, _PMs.ids(pm, nw, :storage), ps)
 end
 
 
 "Create variables for `reactive` storage injection"
-function variable_mc_on_off_storage_reactive(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw)
+function variable_mc_on_off_storage_reactive(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, report::Bool=true)
+    cnds = _PMs.conductor_ids(pm; nw=nw)
```
Comment (sanderclaeys) [inline src/core/variable.jl]: cnds is not used

File: src/core/variable.jl [hunk 4]
```diff
@@ -571,51 +602,63 @@ function variabe_mc_on_off_storage_active(pm::_PMs.AbstractPowerModel; nw::Int=p
...
     ) for i in _PMs.ids(pm, nw, :storage))
+
+    report && _PMs.sol_component_value(pm, nw, :storage, :qs, _PMs.ids(pm, nw, :storage), qs)
 end
 
 
 "voltage variable magnitude squared (relaxed form)"
-function variable_mc_voltage_magnitude_sqr_on_off(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw)
+function variable_mc_voltage_magnitude_sqr_on_off(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, report::Bool=true)
+    cnds = _PMs.conductor_ids(pm; nw=nw)
```
Comment (sanderclaeys) [inline src/core/variable.jl]: cnds is not used here; if you put it, at least fill it out in the cnds expression

File: src/core/variable.jl [hunk 5]
```diff
@@ -571,51 +602,63 @@ function variabe_mc_on_off_storage_active(pm::_PMs.AbstractPowerModel; nw::Int=p
...
     ncnds = length(_PMs.conductor_ids(pm, nw))
 
-    _PMs.var(pm, nw)[:w] = Dict(i => JuMP.@variable(pm.model,
+    w = _PMs.var(pm, nw)[:w] = Dict(i => JuMP.@variable(pm.model,
         [cnd in 1:ncnds], base_name="$(nw)_w_$(i)",
         lower_bound = 0,
         upper_bound = _PMs.ref(pm, nw, :bus, i, "vmax", cnd)^2,
```
Comment (sanderclaeys) [inline src/core/variable.jl]: `_PMs.ref(pm, nw, :bus, i, "vmax", cnd)` => `_PMs.ref(pm, nw, :bus, i, "vmax")[cnd]`

This is my fault actually, I created a new ref method to support this. When all other unit tests work, I will remove this overloading of ref and I will clean up the code.

File: src/prob/test.jl [hunk 6]
```diff
@@ -81,58 +81,64 @@
...
-        variable_active_storage(pm, cnd=c)
-        variable_reactive_storage(pm, cnd=c)
 
-        variable_branch_flow(pm, cnd=c)
-        variable_dcline_flow(pm, cnd=c)
+        constraint_mc_model_voltage(pm, nw=n)
 
-        constraint_model_voltage(pm, cnd=c)
+        variable_storage_indicator(pm)
+        variable_storage_complementary_indicator(pm)
```
Comment (sanderclaeys) [inline src/prob/test.jl]: what about the energy, charge and discharge variables?

File: src/prob/test.jl [hunk 7]
```diff
@@ -81,58 +81,64 @@
...
-    for c in conductor_ids(pm)
-        variable_voltage(pm, cnd=c)
-
-        variable_generation_on_off(pm, cnd=c)
+    for (n, network) in _PMs.nws(pm)
+        variable_mc_voltage(pm, nw=n)
+        variable_mc_branch_flow(pm, nw=n)
+        variable_mc_transformer_flow(pm, nw=n)
+        variable_mc_generation(pm, nw=n)
+        # variable_generation_indicator(pm)
```
Comment (sanderclaeys) [inline src/prob/test.jl]: why commented out?

File: src/prob/test.jl (hunk 6 above)
Comment (sanderclaeys) [reply src/prob/test.jl]: It seems below you simply used `variable_mc_storage`, does this lump them together?

File: test/multiconductor.jl [hunk 8]
```diff
@@ -2,40 +2,36 @@
 
 "an example of building a multi-phase model in an extention package"
 function build_tp_opf(pm::PMs.AbstractPowerModel)
```
Comment (sanderclaeys) [inline test/multiconductor.jl]: This indeed looks like a useless test now.

File: src/core/variable.jl [hunk 9]
```diff
@@ -462,107 +469,131 @@ Create a dictionary with values of type Any for the load.
...
+    else
+        z_gen = _PMs.var(pm, nw)[:z_gen] = JuMP.@variable(pm.model,
+            [i in _PMs.ids(pm, nw, :gen)], base_name="$(nw)_z_gen",
+            lower_bound = 0,
+            upper_bound = 1,
+            start = comp_start_value(ref(pm, nw, :gen, i), "z_gen_start", 1, 1.0)
+        )
+    end
+
+    report && _PMs.sol_component_value(pm, nw, :gen, :gen_status, _PMs.ids(pm, nw, :gen), z_gen)
```
[the remark this answers is not in the document]
Comment (frederikgeth) [reply src/core/variable.jl]: yeah, sadly this is correct (I think)

File: src/core/variable.jl [hunk 10]
```diff
@@ -439,11 +441,14 @@ end
...
     # when extending to 4-wire, this should iterate only over the phase conductors
+    cnds = _PMs.conductor_ids(pm; nw=nw)
+    ncnds = length(_PMs.conductor_ids(pm, nw))
+
     nph = 3
     p_oltc_ids = [id for (id,trans) in _PMs.ref(pm, nw, :transformer) if !all(trans["fixed"])]
-    _PMs.var(pm, nw)[:tap] = Dict(i => JuMP.@variable(pm.model,
+    tap = _PMs.var(pm, nw)[:tap] = Dict(i => JuMP.@variable(pm.model,
         [p in 1:nph],
         base_name="$(nw)_$(p)_tm",
```
[the remark this answers is not in the document]
Comment (frederikgeth) [reply src/core/variable.jl]: thanks!

File: src/core/variable.jl [hunk 11]
```diff
@@ -337,22 +331,30 @@ end
...
 ""
-function variable_mc_active_bus_power_slack(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw)
+function variable_mc_active_bus_power_slack(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, report::Bool=true)
+    cnds = _PMs.conductor_ids(pm; nw=nw)
     ncnds = length(_PMs.conductor_ids(pm, nw))
 
-    _PMs.var(pm, nw)[:p_slack] = Dict(i => JuMP.@variable(pm.model,
-        [cnd in 1:cnds], base_name="$(nw)_$(cnd)_p_slack",
+    p_slack = _PMs.var(pm, nw)[:p_slack] = Dict(i => JuMP.@variable(pm.model,
+        [cnd in 1:cnds], base_name="$(nw)_p_slack_$(cnd)",
```
[the remark this answers is not in the document]
Comment (frederikgeth) [reply src/core/variable.jl]: thanks!

File: src/core/variable.jl [hunk 12]
```diff
@@ -337,22 +331,30 @@ end
...
 ""
-function variable_mc_reactive_bus_power_slack(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw)
-    _PMs.var(pm, nw, cnd)[:q_slack] = JuMP.@variable(pm.model,
-        [i in _PMs.ids(pm, nw, :bus)], base_name="$(nw)_$(cnd)_q_slack",
+function variable_mc_reactive_bus_power_slack(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, report::Bool=true)
+    cnds = _PMs.conductor_ids(pm; nw=nw)
+    ncnds = length(_PMs.conductor_ids(pm, nw))
+
+    q_slack = _PMs.var(pm, nw)[:q_slack] = Dict(i => JuMP.@variable(pm.model,
+        [cnd in 1:cnds], base_name="$(nw)_q_slack_$(cnd)",
```
[the remark this answers is not in the document]
Comment (frederikgeth) [reply src/core/variable.jl]: thanks!

File: src/core/variable.jl [hunk 13]
```diff
@@ -462,107 +469,131 @@ Create a dictionary with values of type Any for the load.
...
     end
+
+    report && _PMs.sol_component_value(pm, nw, :bus, :status, _PMs.ids(pm, nw, :bus), z_voltage)
 end
 
 
 "Create variables for generator status"
-function variable_mc_indicator_generation(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, relax=false)
-    _PMs.variable_generation_indicator(pm; nw=nw, relax=relax)
+function variable_mc_indicator_generation(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, relax::Bool=false, report::Bool=true)
```
[the remark this answers is not in the document]
Comment (frederikgeth) [reply src/core/variable.jl]: agreed, added it to WIP issues. not going to address this currently

File: src/core/variable.jl (hunk 5 above)
Comment (frederikgeth) [reply src/core/variable.jl]: ok

File: src/prob/test.jl (hunk 6 above)
Comment (frederikgeth) [reply src/prob/test.jl]: they are variants depending on whether it is actually multiperiod or not

File: src/core/variable.jl (hunk 2 above)
Comment (frederikgeth) [reply src/core/variable.jl]: ok

File: src/core/variable.jl (hunk 3 above)
Comment (frederikgeth) [reply src/core/variable.jl]: ok

File: src/core/variable.jl (hunk 4 above)
Comment (frederikgeth) [reply src/core/variable.jl]: ok

File: src/prob/test.jl (hunk 7 above)
Comment (frederikgeth) [reply src/prob/test.jl]: removed now

**Comment by frederikgeth** [comment]
959 passed, 103 failed, 61 errored, 0 broken

merging now

**Pull request merged**